From 1647449b63d00c562eb85a78681617ac55a2cfe4 Mon Sep 17 00:00:00 2001
From: Bai Ping <ping.bai@nxp.com>
Date: Fri, 9 Jun 2017 18:10:35 +0800
Subject: [PATCH 1883/5242] MLK-15056 arm: imx: use static mapping for imx7ulp
 AIPSx memory space

commit  be9956fdf553beaed6b47023bb984199dcd30810 from
https://source.codeaurora.org/external/imx/linux-imx.git

The AIPSx address space of i.MX7ULP need to be mapped as SZ_1M block
in iRAM tlb for suspend code use. If we use ioremap to map these
address region into kernel space, we can't make sure that the returned
virtual address is 1M alignment. So we can map this address regions
as static, then if we use the ioremap to map these memory regions, it will
always return the virtual address of static mapping. So we can make sure
the virtual address is 1M aligned.

Signed-off-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/mach-imx7ulp.c |   10 ++++++++++
 arch/arm/mach-imx/mx7ulp.h       |   11 +++++++++++
 2 files changed, 21 insertions(+)

diff --git a/arch/arm/mach-imx/mach-imx7ulp.c b/arch/arm/mach-imx/mach-imx7ulp.c
index ff70624..2d916e1 100644
--- a/arch/arm/mach-imx/mach-imx7ulp.c
+++ b/arch/arm/mach-imx/mach-imx7ulp.c
@@ -16,6 +16,15 @@
 #include "cpuidle.h"
 #include "hardware.h"
 
+/* static IO mapping, and ioremap() could always share the same mapping. */
+static struct map_desc mx7ulp_io_desc[] __initdata = {
+	mx7ulp_aips_map_entry(1, MT_DEVICE),
+	mx7ulp_aips_map_entry(2, MT_DEVICE),
+	mx7ulp_aips_map_entry(3, MT_DEVICE),
+	mx7ulp_aips_map_entry(4, MT_DEVICE),
+	mx7ulp_aips_map_entry(5, MT_DEVICE),
+};
+
 static void __init imx7ulp_init_machine(void)
 {
 	struct device *parent;
@@ -38,6 +47,7 @@ static void __init imx7ulp_init_irq(void)
 
 static void __init imx7ulp_map_io(void)
 {
+	iotable_init(mx7ulp_io_desc, ARRAY_SIZE(mx7ulp_io_desc));
 	imx7ulp_pm_map_io();
 }
 
diff --git a/arch/arm/mach-imx/mx7ulp.h b/arch/arm/mach-imx/mx7ulp.h
index a5f9bb27..a00e86f 100644
--- a/arch/arm/mach-imx/mx7ulp.h
+++ b/arch/arm/mach-imx/mx7ulp.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2016 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright NXP 2017.
  */
 
 /*
@@ -45,6 +46,16 @@
 #define MX7ULP_MMDC_IO_BASE_ADDR	0x40ad0000
 #define MX7ULP_MMDC_IO_SIZE		0x1000
 
+/* below is just used for static mapping of the AIPSx's memory region */
+#define MX7ULP_AIPS_VIRT_BASE(x)	(0xf4000000 + ((x) * SZ_1M))
+
+#define mx7ulp_aips_map_entry(index, _type) {				\
+	.virtual = MX7ULP_AIPS_VIRT_BASE(index),			\
+	.pfn = __phys_to_pfn(MX7ULP_AIPS ## index ## _BASE_ADDR),	\
+	.length	= SZ_1M,						\
+	.type = _type,							\
+}
+
 #define TT_ATTRIB_NON_CACHEABLE_1M	0x802
 #define MX7ULP_IRAM_TLB_SIZE		0x4000
 #define MX7ULP_SUSPEND_OCRAM_SIZE	0x1000
-- 
1.7.9.5

